src/envoy/http/authn/test_utils.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "common/protobuf/protobuf.h"
+#include "gmock/gmock.h"
+#include "src/envoy/http/authn/context.pb.h"
+
+namespace Envoy {
+namespace Http {
+namespace IstioAuthN {
+
+std::unique_ptr<Payload> CreateX509Payload(const std::string& user) {
+  std::unique_ptr<Payload> payload(new Payload);
+  payload->mutable_x509()->set_user(user);
+  return payload;
+}
+
+std::unique_ptr<Payload> CreateJwtPayload(const std::string& user,
+                                          const std::string& presenter) {
+  std::unique_ptr<Payload> payload(new Payload);
+  payload->mutable_jwt()->set_user(user);
+  if (!presenter.empty()) {
+    payload->mutable_jwt()->set_presenter(presenter);
+  }
+  return payload;
+}
+
+Result AuthNResultFromString(const std::string& text) {
+  Result result;
+  EXPECT_TRUE(Protobuf::TextFormat::ParseFromString(text, &result));
+  return result;
```

Comment:
You copy the whole proto when returning.  That is not good.  Even it is testing code, but I hate to see such c++ code.

Reply:
:), but you don't want pointer neither.
This function is mainly to reduce the hassle of having multiple lines to construct expected proto to be use in test proto equal. It's for test, also, modern compiler can also optimize this, so I like to keep it. I change the namespace to make it more standout for testing.

Reply:
In that case, it is prefer to return const T&

Reply:
Nah, should not return ref to a local variable.

Btw, https://en.wikipedia.org/wiki/Copy_elision (skip to return value optimization) explains returning by value is no longer inefficient with modern compiler (gcc is one).